src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/FeatureSwitchDataFlow.cs

```diff
@@ -0,0 +1,995 @@
+					RequiresDynamicCode ();
+			}
+
+			static void GuardedElse ()
+			{
+				if (!RuntimeFeature.IsDynamicCodeSupported)
+				{
+					throw new Exception ();
+				}
+				else
+				{
+					RequiresDynamicCode ();
+				}
+			}
+
+			static void GuardedAnd ()
+			{
+				var a = RuntimeFeature.IsDynamicCodeSupported && RequiresDynamicCodeBool ();
+			}
+
+			static void GuardedOr ()
+			{
+				var a = !RuntimeFeature.IsDynamicCodeSupported || RequiresDynamicCodeBool ();
+			}
+
+			static void GuardedTernary ()
+			{
+				var a = RuntimeFeature.IsDynamicCodeSupported ? RequiresDynamicCodeBool () : true;
+				var b = !RuntimeFeature.IsDynamicCodeSupported ? true : RequiresDynamicCodeBool ();
+			}
+
+			static void GuardedThrow ()
+			{
+				if (!RuntimeFeature.IsDynamicCodeSupported)
+				{
+					throw new Exception ();
+				}
+
+				RequiresDynamicCode ();
+			}
+
+			// NativeAot doesn't optimize branches away based on DoesNotReturnIfAttribute
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void GuardedAssert ()
+			{
+				Debug.Assert (RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+
+			// NativeAot doesn't optimize branches away based on DoesNotReturnIfAttribute
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void GuardedDoesNotReturnIfTrue ()
+			{
+				DoesNotReturnIfTrue (!RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+
+			// NativeAot doesn't optimize branches away based on DoesNotReturnIfAttribute
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void GuardedDoesNotReturnIfFalse ()
+			{
+				DoesNotReturnIfFalse (RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+
+			// NativeAot doesn't optimize branches away based on DoesNotReturnAttribute
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void GuardedDoesNotReturn ()
+			{
+				if (!RuntimeFeature.IsDynamicCodeSupported)
+					DoesNotReturn ();
+
+				RequiresDynamicCode ();
+			}
+
+			// NativeAot doesn't optimize branches away based on DoesNotReturnIfAttribute
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void GuardedDoesNotReturnIfFalseCtor ()
+			{
+				new DoesNotReturnIfFalseCtor (RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+		}
+
+		class FeatureCheckBooleanExpressions
+		{
+			[ExpectedWarning ("IL2026", nameof (RequiresUnreferencedCode), ProducedBy = Tool.Trimmer | Tool.NativeAot)]
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void And ()
+			{
+				if (RuntimeFeature.IsDynamicCodeSupported && TestFeatures.IsUnreferencedCodeSupported) {
+					RequiresDynamicCode ();
+					RequiresUnreferencedCode ();
+				}
```

Comment:
How come NativeAOT produces 3050 here? I would expect the branch to be eliminated. I must be missing something here. Is it because of compound condition?